Improve Preparation form links to Interactions

grantfitzsimmons opened the pull request:
Fixes #6112

### Checklist

- [X] Self-review the PR after opening it to make sure the changes look good and
      self-explanatory (or properly documented)
- [ ] Add relevant documentation (Tester - Dev)

#### Before

You could not see any linked disposals or exchanges. Interactions that were empty still displayed with no results.

<img width="1191" alt="image" src="https://github.com/user-attachments/assets/f9f780c2-8548-4029-ac61-c3a453329396" />

#### After

* Interactions are only shown when they are linked to a Preparation
	* If none are linked, it shows "There are no interactions linked to this preparation"
* When clicking on the "Show Interactions" button, the user should see all associated
   * Resolved Loans
   * Unresolved Loans
   * Gifts
   * Disposals
   * Exchange Out
   * Exchange In

https://github.com/user-attachments/assets/4159042e-f0af-417f-ac6e-309dd1f77da8


<!-- What are the steps to verify the fixes/changes in this PR? -->
<!-- Can part of that be replaced by automated tests? -->

### Testing instructions

You will need to make sure you have the "Show Interactions" / "Show Loans" button on the Preparation form to test this properly. If you do not already, here is the XML to add:

```xml
 <cell type="command" id="ShowLoansBtn" name="ShowLoansBtn" label="SHOW_INTERACTIONS" ignore="true"/>
```
- [ ] Create a new CO with a memorable catalog number (you'll need it)
**No Interactions:**
- [ ] Create a new Preparation, give it a large count, and save
- [ ] Click on the 'Show Interactions' button and see that "There are no interactions linked to this preparation." displays
**Gift:**
- [ ] Create a new Loan using the catalog number
- [ ] Select some portion of your preparation's available count and create the loan
- [ ] Go back to the Preparation and click on the 'Show Interactions' button– see it appears under "Open Loans"
- [ ] Resolve the loan
- [ ] Go back to the Preparation and click on the 'Show Interactions' button– see it appears under "Resolved Loans"
**Gift:**
- [ ] Create a new Gift using the catalog number
- [ ] Select some portion of your preparation's available count and create the gift
- [ ] Go back to the Gift and click on the 'Show Interactions' button– see it appears under "Gift"
**Exchange Out:**
- [ ] Create a new Exchange Out using the catalog number
- [ ] Select some portion of your preparation's available count and create the Exchange Out
- [ ] Go back to the Exchange Out and click on the 'Show Interactions' button– see it appears under "Exchange Out"
**Exchange In:**
- [ ] Create a new Exchange In using the catalog number
- [ ] Select some portion of your preparation's available count and create the Exchange In
- [ ] Go back to the Exchange In and click on the 'Show Interactions' button– see it appears under "Exchange In"
**Disposal:**
- [ ] Create a new Disposal using the catalog number
- [ ] Select some portion of your preparation's available count and create the Disposal
- [ ] Go back to the Disposal and click on the 'Show Interactions' button– see it appears under "Disposal"

Test this by using a combination of various interactions and preparations. Involve the preparation in multiple interactions of the same type. Verify that existing preparations linked to interactions appear as expected. Compare the behavior to production and let me know if this is undesirable.
- [ ] OK I did it

If anyone is interested in updating the documentation for this, I would appreciate it. Thank you!

grantfitzsimmons committed:
Update Interaction localization

grantfitzsimmons committed:
Add Disposal to Interactions dialog

grantfitzsimmons committed:
Update interactions.ts

grantfitzsimmons committed:
Hide interactions when none are available

grantfitzsimmons committed:
Remove unused text

grantfitzsimmons committed:
Use localized Preparation table name

grantfitzsimmons committed:
Lint code with ESLint and Prettier

Triggered by 81067b54830456c54c3e1e6fe8d30fc81739ba08 on branch refs/heads/issue-6108

grantfitzsimmons committed:
Improve localization for Preparations

grantfitzsimmons committed:
Resolve conflicts

grantfitzsimmons committed:
Update interactions.ts

grantfitzsimmons committed:
Fix no interactions label

grantfitzsimmons committed:
Improve Preparation localization further

grantfitzsimmons committed:
Finish localizing preparations

grantfitzsimmons committed:
Merge branch 'production' into issue-6108

grantfitzsimmons committed:
Add tables import

grantfitzsimmons committed:
Lint code with ESLint and Prettier

Triggered by 59802e1475245a3c6b8fc610ade9fea2677dcd7b on branch refs/heads/issue-6108

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/FormCommands/index.tsx:
```diff
@@ -98,7 +99,9 @@ const commandRenderers: {
             >
-              {interactionsText.preparationsCanNotBeReturned()}
+              {interactionsText.preparationsCanNotBeReturned({
+                preparationTable: String(tables.Preparation.label).toLowerCase(),
```
String() is probably not needed here and in the following usages

```suggestion
                preparationTable: tables.Preparation.label.toLowerCase(),
```

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/Interactions/PrepDialog.tsx:
```diff
@@ -133,7 +136,10 @@ export function PrepDialog({
       }
-      header={interactionsText.preparations()}
+      header=
+          {interactionsText.preparations({
+            preparationTable: tables.Preparation.label,
+        })}
```
interesting that prettier did not reformat this file in https://github.com/specify/specify7/pull/6110/commits/0371f027fa8349965873396f247440e92828b18d

part of https://github.com/specify/specify7/issues/6051 I guess

melton-jason committed:
Show all related interactions in ShowLoans dialog

Alternative implementation of #6110 for #6108

melton-jason committed:
Remove duplicates from returned Interaction request

melton-jason commented on specifyweb/frontend/js_src/lib/components/FormCommands/ShowTransactions.tsx:
```diff
@@ -100,74 +100,149 @@ export function ShowLoansCommand({
                 domainFilter: false,
               }).then(({ records }) => records.map(deserializeResource))
             : undefined,
+          exchangeIns: hasTablePermission('ExchangeInPrep', 'read')
+            ? fetchCollection('ExchangeInPrep', {
+                limit: DEFAULT_FETCH_LIMIT,
+                preparation: preparation.get('id'),
+                domainFilter: false,
+              }).then(({ records }) => records.map(deserializeResource))
+            : undefined,
```
I have actually worked on a branch - [issue-6108-b](https://github.com/specify/specify7/compare/issue-6108...issue-6108-b) - to address this problem. It is functionally a complete reimplementation of the Show Transactions Dialog: take a look through the branch and see if there's any ideas you'd like to carry over to this implementation!

Here are some videos comparing production, this branch, and issue-6108-b: 

**production**

https://github.com/user-attachments/assets/2d033a52-a59e-4fe1-b3a9-0ce5f5e1ba01

_The error here is a bug in production which occurs whenever there are ExchangeOut records associated with a Preparation_

**issue-6108**

https://github.com/user-attachments/assets/0490c8c3-03c3-4079-843c-1df7c265e030


**issue-6108-b**

https://github.com/user-attachments/assets/68e9b249-8e3e-4f6e-84d3-f9733f998036

grantfitzsimmons committed:
Update specifyweb/frontend/js_src/lib/components/FormCommands/index.tsx

Co-authored-by: Max Patiiuk <max@patii.uk>

grantfitzsimmons commented on specifyweb/frontend/js_src/lib/components/FormCommands/ShowTransactions.tsx:
```diff
@@ -100,74 +100,149 @@ export function ShowLoansCommand({
+        </>
+      ) : (
+        <>
+          {Array.isArray(data.openLoans) && data.openLoans.length > 0 && (
```
This was intentional!

melton-jason committed:
Merge remote-tracking branch 'origin/main' into issue-6108-b

melton-jason committed:
Run Prettier on all files

melton-jason committed:
Merge remote-tracking branch 'origin/issue-6108-b' into issue-6108

melton-jason committed:
Remove unused localization strings

melton-jason committed:
Move handling of fetching related Interactions to parent

melton-jason commented:
I've updated the branch, resolving merge conflicts and merged my changes from `issue-6108-b` (see https://github.com/specify/specify7/pull/6110#discussion_r1922088752)! 

https://github.com/user-attachments/assets/b2aff00f-bd23-4992-8885-e75ebcfe2dbf

This is ready for testing as soon as the testing instructions are updated! 🚀

melton-jason committed:
Lint code with ESLint and Prettier

Triggered by aadcb21bfa480ea8b68eb73470eca323b61d32ee on branch refs/heads/issue-6108

grantfitzsimmons committed:
Merge branch 'main' into issue-6108

grantfitzsimmons committed:
fix(localization): simplify preparation labels

grantfitzsimmons committed:
fix(tests): misc fixes

grantfitzsimmons committed:
fix(localization): remove unused strings

grantfitzsimmons commented:
@melton-jason Even when I do `git checkout e1a61345410416f65fe5c5d7832ad0d1c01c47e1`, I don't see what you show in your video. Could you check it out (when we have time near 7.13)?

bhumikaguptaa reviewed (changes requested):
- [X] Create a new CO with a memorable catalog number (you'll need it)
**No Interactions:**
- [X] Create a new Preparation, give it a large count, and save
- [X] Click on the 'Show Interactions' button and see that "There are no interactions linked to this preparation." displays
**Gift:**
- [X] Create a new Loan using the catalog number
- [X] Select some portion of your preparation's available count and create the loan
- [X] Go back to the Preparation and click on the 'Show Interactions' button– see it appears under "Open Loans."
- [ ] Resolve the loan
- [ ] Go back to the Preparation and click on the 'Show Interactions' button– see it appears under "Resolved Loans."

<img width="1458" height="806" alt="Image" src="https://github.com/user-attachments/assets/bd3a2301-9b17-4329-895b-75e68fd3ce32" />

Every time I try to return a loan in the pop-up window instead of opening the form completely, the screen freezes. 

<img width="961" height="447" alt="Image" src="https://github.com/user-attachments/assets/c5a92045-f7cb-4ea3-8dea-6135144a74e5" />

Secondly, the resolved loan records do not appear under "Resolved loans" as a separate category, they are just a part of the Loans overall.

**Gift:**
- [X] Create a new Gift using the catalog number
- [X] Select some portion of your preparation's available count and create the gift
- [X] Go back to the Gift and click on the 'Show Interactions' button– see it appears under "Gift"
**Exchange Out:**
- [X] Create a new Exchange Out using the catalog number
- [X] Select some portion of your preparation's available count and create the Exchange Out
- [X] Go back to the Exchange Out and click on the 'Show Interactions' button– see it appears under "Exchange Out"
**Exchange In:**
- [X] Create a new Exchange In using the catalog number
- [X] Select some portion of your preparation's available count and create the Exchange In
- [X] Go back to the Exchange In and click on the 'Show Interactions' button– see it appears under "Exchange In"
**Disposal:**
- [X] Create a new Disposal using the catalog number
- [X] Select some portion of your preparation's available count and create the Disposal
- [X] Go back to the Disposal and click on the 'Show Interactions' button– see it appears under "Disposal"

Test this by using a combination of various interactions and preparations. Involve the preparation in multiple interactions of the same type. Verify that existing preparations linked to interactions appear as expected. Compare the behavior to production and let me know if this is undesirable.
- [X] OK I did it

Overall everything works except the loan features. 
Link to DB: https://ojsmnh20250909-issue-6108.test.specifysystems.org/specify/view/collectionobject/454948/ 
Catalog number used in DB: 000005037

Commit:
Merge remote-tracking branch 'origin/main' into issue-6108

Commit:
Lint code with ESLint and Prettier

Triggered by f6fa501078674a8ef8ef38b09e0dd9fe8db60509 on branch refs/heads/issue-6108

bhumikaguptaa reviewed (approved):
- [X] Create a new CO with a memorable catalog number (you'll need it)
**No Interactions:**
- [X] Create a new Preparation, give it a large count, and save
- [X] Click on the 'Show Interactions' button and see that "There are no interactions linked to this preparation." displays
**Gift:**
- [X] Create a new Loan using the catalog number
- [X] Select some portion of your preparation's available count and create the loan
- [ X] Go back to the Preparation and click on the 'Show Interactions' button– see it appears under "Open Loans"
- [X] Resolve the loan
- [X] Go back to the Preparation and click on the 'Show Interactions' button– see it appears under "Resolved Loans"
**Gift:**
- [X] Create a new Gift using the catalog number
- [x] Select some portion of your preparation's available count and create the gift
- [x] Go back to the Gift and click on the 'Show Interactions' button– see it appears under "Gift"
**Exchange Out:**
- [X] Create a new Exchange Out using the catalog number
- [X] Select some portion of your preparation's available count and create the Exchange Out
- [X] Go back to the Exchange Out and click on the 'Show Interactions' button– see it appears under "Exchange Out"
**Exchange In:**
- [X] Create a new Exchange In using the catalog number
- [X] Select some portion of your preparation's available count and create the Exchange In
- [X] Go back to the Exchange In and click on the 'Show Interactions' button– see it appears under "Exchange In"
**Disposal:**
- [X] Create a new Disposal using the catalog number
- [X] Select some portion of your preparation's available count and create the Disposal
- [X] Go back to the Disposal and click on the 'Show Interactions' button– see it appears under "Disposal"

Test this by using a combination of various interactions and preparations. Involve the preparation in multiple interactions of the same type. Verify that existing preparations linked to interactions appear as expected. Compare the behavior to production and let me know if this is undesirable.
- [X] OK I did it

--
Everything mentioned in the checklist is working and looks good! Unsure what caused for the page to freeze last time. 

--
I noticed an error unrelated to the checklist that is not getting recreated in 7.11.4 but I get an error message on the Branch. 



https://github.com/user-attachments/assets/7430ed4a-3fd4-4bc7-9cff-389176aedd0e



Error message: [Specify 7 Crash Report - 2026-03-17T14_29_56.750Z.txt](https://github.com/user-attachments/files/26064995/Specify.7.Crash.Report.-.2026-03-17T14_29_56.750Z.txt)
Link: https://ojsmnh20251211-issue-6108.test.specifysystems.org/specify/overlay/interactions/create/Loan/

acwhite211 committed:
Merge branch 'main' into issue-6108

acwhite211 committed:
Lint code with ESLint and Prettier

Triggered by 9bf1dd3222b91907d246503b45739871f0db26d7 on branch refs/heads/issue-6108

Commit:
Merge remote-tracking branch 'origin/main' into issue-6108

Commit:
Lint code with ESLint and Prettier

Triggered by 7a814e893f42659f86f7918ada00fa726da625b5 on branch refs/heads/issue-6108

CarolineDenis committed:
Merge branch 'main' into issue-6108

acwhite211 committed:
Merge branch 'main' into issue-6108

acwhite211 committed:
Lint code with ESLint and Prettier

Triggered by a6a8452a6009ebe42c3b76bae54bcbeb45ecb344 on branch refs/heads/issue-6108
